Add a section on 'deployment' to documentation

tanhaow opened the pull request:
This PR is to resolve issue #279.

laurejt reviewed (commented):
The deployment sections look pretty good, although there are a few places where instructions should be expanded.

I think there's a higher level question of what this document is for. I thought it was for a guide for ansible-specific deployment, but this also includes a guide for CDH's general release process. It seems to me that this more general information should live somewhere else like the RSE Handbook.

Redundancy isn't necessarily a bad thing; but we'd need to update this any time we update the release checklist or our release workflow more generally. @rlskoeser, what do you think?

laurejt commented on docs/Release & Deployment Guide.md:
That said, I really like the thorough release documentation; just want to make sure it's located in the right place :)

laurejt reviewed (commented):
Here's a series of comments on the release sections. In the future, the PR should describe the PR in more detail. It wasn't obvious to me what the purpose of this document was since it does not match the issue description.

rlskoeser reviewed (changes requested):
@tanhaow thank you for this detailed, thoughtful work helping to start articulate and document our processes. I agree with @laurejt, there are questions of scope and redundancy for this documentation.  I suspect the main problem is that you trying to do too much.  (I also see that we didn't scope the task at all in the related issue, so it's understandable.)

I propose splitting this out and focus on the deploy process specifically, independent of release prep workflows (which are less clearly related to cdh-ansible).  What I would like it to include:
- how to deploy web apps to staging servers, to production 
   - what the default deploy branch is for each
   - how to deploy a non-default branch
- how to deploy from a local cdh-ansible checkout, how to deploy with Ansible Tower (and note that all playbooks may not be  available in Tower yet)

Some of this is already documented in the main cdh-ansible readme, in the section starting with "Running a Playbook" ; probarbly some of that content could shift from the readme to the new deploy document, and we could link from the readme to the deploy doc. 

It might make sense to do that as a new branch so we can keep this draft open while we decide what to do with the release process documentation.

I'll chime in on some of the conversations on this PR, but for the larger process/workflow documentation we should discuss as a team where and how we want to to document these things.  I agree we should not duplicate what is in the software checklist template; if that checklist is not clear, then let's improve it and make it more useful for everyone (e.g., more clearly flagging optional steps that don't apply to all projects).

rlskoeser commented on docs/Release & Deployment Guide.md:
```diff
@@ -0,0 +1,165 @@
+
+5. Check `DEPLOYNOTES` to see if there's any commands you need to manually run on sever.
+
+    - Find server hostnames in the [CDH ansible host doc](https://github.com/Princeton-CDH/cdh-ansible/blob/main/inventory/all_hosts). Pay attention to group names (e.g. cdhweb_staging, cdhweb_production), which specify staging vs production servers.
```
also recommend naming the file, since this should be accessible locally as well and is good for people to be conversant with the inventory file; use a relative link if possible

rlskoeser commented on docs/Release & Deployment Guide.md:
````diff
@@ -0,0 +1,165 @@
+   ```
+   > **Why?** Git-flow automatically creates `release/3.15` from `develop`. This branch is for final fixes (version bumps, changelog updates, small bug fixes).
+
+4. It's wise to publish the release branch so others can contribute:
+
+   ```bash
+   git flow release publish 3.15
+   ```
````
I wasn't familiar with the `git flow release publish` command but I'm guessing it is equivalent to pushing the release branch. We do sometimes collaborate on releases, at least for review. It's also useful to create a GitHub PR to run automated checks before merging to main.

rlskoeser commented:
@tanhaow when you revise, please use lower case and dashes for the documentation filename, and add it to the readme in the docs directory.  (We might want to move the instructions on PUL VM access to a separate documentation file, but let's do that separately.)

tanhaow commented on docs/Release & Deployment Guide.md:
```diff
@@ -0,0 +1,165 @@
+    3. Enter your release branch name (e.g., `release/3.15`).
+    4. Go to **Next**, preview, and **Launch**.
+
+5. Check `DEPLOYNOTES` to see if there's any commands you need to manually run on sever.
```
Do I need to include an explanation on why there might be more than one machine - or is a simple mention enough?

rlskoeser replied:
Brief mention here sounds appropriate - if helpful maybe we could explain in more detail why that is and what kinds of impact it might have, and then this text could add a link to that documentation.

Maybe just say that typically we have two VMs for each application in each environment (or whatever wording conveys that sensibly)

tanhaow commented:
Thank you @laurejt and @rlskoeser for the detailed review feedback! I've made changes based on your suggestions:
Improvements made:
- Added scope clarification at the beginning explaining what this guide covers
- Made deployment instructions more explicit (template navigation, page-by-page steps, do not click on the rocket button 🚀😂, etc.)
- Clarified conditional steps for JavaScript/Python projects
- Updated changelog section to reflect current conventions (review vs. create)
- Improved readability by replacing faded blockquote styling with clearer formatting
- Renamed file to release-and-deployment-guide.md and added to docs README

Note:
@rlskoeser's suggestion to split up the documentation for release and deployment workflows makes a lot of sense. I've create a new issue (#286) in ice box to get back when we have capacity.

rlskoeser reviewed (changes requested):
Thanks for your revisions; sorry you were not clear on my requested changes or the scope for this documentation.

Please revise so this only includes deployment and not release preparation, and covers all important aspects of deployment.

The deploy documentation should include:
- how to deploy web apps to staging servers, to production
  - the default deploy branch is for staging and production deploy
    -  how to deploy a non-default branch
- how to deploy from a local cdh-ansible checkout (command-line deploy)
- how to deploy with Ansible Tower 
    - add a note that all playbooks are not yet available in Tower 

If you have any questions on these, first consult the main cdh-ansible readme; if anything is still unanswered, make a list of things where you need my help.

If there is important release preparation documentation that is not currently covered by the team handbook or the software release checklist issue template, please suggest changes to the team handbook with the documentation you'd like to have included. We can track that as a separate task if that's helpful.

rlskoeser commented on docs/README.md:
```diff
@@ -5,6 +5,10 @@
 - [Prodigy](applications/prodigy.md)
 - [Princeton Geniza Project](applications/geniza.md)
 
+## Deployment and Infrastructure
+
+- [Release and Deployment Guide](release-and-deployment-guide.md) - Complete workflow for releasing and deploying CDH applications using Ansible
+
```
Please update title, link, and description to match the revised document

tanhaow replied:
Thank you for catching this!!

tanhaow replied:
I felt the description is a bit redundant so I removed it the new commit

rlskoeser commented on docs/deployment-guide.md:
````diff
@@ -0,0 +1,109 @@
+ansible-playbook playbooks/your_app.yml --tags=all,final-pause
+```
+
+**Revert a Deployment:**
````
I'm not 100% confident in this playbook, some kind of warning is probably good to include. How does this sound to you?

```suggestion
**Revert a Deployment:**
**⚠️ Note:** Does not reverse django migrations; this playbook only switches current and previous symlinks.
```

tanhaow replied:
This is great to add.

rlskoeser reviewed (changes requested):
This revision looks great. Just needs an update to the link in the main docs readme file and then I think it will be ready to merge.
